### PR TITLE
Z-Mimic

### PR DESCRIPTION
**Pull request by Lohikar**
Per request/commission of Kapu, a port of Z-Mimic.

---

This is an alternative to the turf transparency element / vis_contents. While it is internally more complex, it should provide higher quality render + the ability to monkeypatch atoms as they are being copied. Support for wide and long atoms is included in this revision, controlled by `ZMM_LOOKAHEAD`/`ZMM_LOOKBESIDE`/`ZMM_WIDE_LOAD`.

On other codebases, this system has been tested to scale up to at least 150k z-turfs.

Limitations:
- Z-Mimic (currently) cannot copy vis_contents.
- Atoms copied by zmimic must still make sense without their plane var.
- For appearance updates that aren't overlay updates, `UPDATE_OO_IF_PRESENT` must be used to notify ZM of changes.
- This revision of ZM only supports linear z-level order. Skipping z-levels within a z-stack is not supported and will result in layering corruption, though this is not a fundamental limitation.

todo:
- [x] Lighting update propagation is broken.
- [x] Z-Copy doesn't have priority set.
- [x] Symbols referring to 'OO' should be renamed.
- [x] ~~Overlays hook is missing.~~ Replaced with `update_appearance()` hook.
- [x] Z-Copy needs to be updated to properly use helpers like `suspend()`.
- [x] Docs are out of date.
- [x] The movable destruction hook is missing.

## Changelog

:cl:
refactor: Z-holes are now rendered using Z-Mimic, which should improve accuracy and performance.
/:cl:

### CODE REVIEW

File: code/modules/zmimic/mimic_movable.dm [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+	zmm_flags = ZMM_IGNORE  // Only one of these should ever be visible at a time, the mimic logic will handle that.
+
+/atom/movable/openspace/turf_proxy/attackby(obj/item/W, mob/user)
+	loc.attackby(W, user)
+
+/atom/movable/openspace/turf_proxy/attack_hand(mob/user as mob)
+	loc.attack_hand(user)
+
+/atom/movable/openspace/turf_proxy/attack_generic(mob/user as mob)
+	loc.attack_generic(user)
```
Comment (Kapu1178) [inline code/modules/zmimic/mimic_movable.dm:211]: This is likely necessary to not fuck the attack chain
```suggestion
/atom/movable/openspace/turf_proxy/attackby(obj/item/W, mob/user)
	return loc.attackby(W, user)

/atom/movable/openspace/turf_proxy/attack_hand(mob/user as mob)
	return loc.attack_hand(user)

/atom/movable/openspace/turf_proxy/attack_generic(mob/user as mob)
	return loc.attack_generic(user)
```

File: code/modules/zmimic/mimic_movable.dm [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+/atom/movable/openspace/turf_mimic/Initialize(mapload, ...)
+	. = ..()
+	ASSERT(isturf(loc))
+	delegate = loc:below
+
+/atom/movable/openspace/turf_mimic/attackby(obj/item/W, mob/user)
+	loc.attackby(W, user)
+
+/atom/movable/openspace/turf_mimic/attack_hand(mob/user as mob)
+	to_chat(user, span_notice("You cannot reach \the [src] from here."))
```
Comment (Kapu1178) [inline code/modules/zmimic/mimic_movable.dm:235]: sam,e here
```suggestion
/atom/movable/openspace/turf_mimic/attackby(obj/item/W, mob/user)
	return loc.attackby(W, user)

/atom/movable/openspace/turf_mimic/attack_hand(mob/user as mob)
	to_chat(user, span_notice("You cannot reach \the [src] from here."))
	return FALSE
```

**Review by Kapu1178 — CHANGES_REQUESTED** [review]

File: code/__HELPERS/cmp.dm [hunk 3]
```diff
@@ -175,3 +175,7 @@ GLOBAL_VAR_INIT(cmp_field, "name")
 ///Orders R-UST fusion by priority
 /proc/cmp_fusion_reaction_des(datum/fusion_reaction/A, datum/fusion_reaction/B)
 	return B.priority - A.priority
+
+/// Sort by plane, then by layer. Approximately BYOND rendering order.
+/proc/cmp_zm_render_order(atom/A, atom/B)
+	return (B.plane - A.plane) || (B.layer - A.layer)
```
Comment (itsmeow) [inline code/__HELPERS/cmp.dm:181]: this is identical to `cmp_atom_layer_asc`

File: code/modules/zmimic/mimic_movable.dm [hunk 4]
```diff
@@ -9,6 +9,10 @@
 	if (. && bound_overlay)
 		bound_overlay.setDir(ndir)
 
+/atom/movable/update_appearance(updates)
+	. = ..()
+	UPDATE_OO_IF_PRESENT
```
Comment (itsmeow) [inline code/modules/zmimic/mimic_movable.dm:14]: this is already in mimic_common

Comment (Lohikar) [reply code/modules/zmimic/mimic_movable.dm:14]: Yes, I know; wasn't paying attention.

File: code/__DEFINES/flags.dm [hunk 5]
```diff
@@ -289,12 +289,41 @@ GLOBAL_LIST_INIT(bitflags, list(1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 204
...
+	"Z_ATMOS_IN_DOWN",
+	"Z_ATMOS_OUT_UP",
+	"Z_ATMOS_OUT_DOWN",
+
+	"Z_MIMIC_BELOW",
+	"Z_MIMIC_OVERWRITE",
+	"Z_MIMIC_NO_AO",
+	"Z_MIMIC_NO_OCCLUDE",
+	"Z_MIMIC_BASETURF"
+))
```
Comment (Kapu1178) [inline code/__DEFINES/flags.dm:316]: You're probably looking for bitfields.dm

File: code/game/atoms.dm [hunk 6]
```diff
@@ -772,6 +776,9 @@
 	if(updates & UPDATE_ICON)
 		. |= update_icon(updates)
 
+	if (ismovable(src))
+		UPDATE_OO_IF_PRESENT
```
Comment (Kapu1178) [inline code/game/atoms.dm:780]:
```suggestion
	// This is not an override for performance reasons.
	if (ismovable(src))
		UPDATE_OO_IF_PRESENT
```

File: code/game/atoms_movable.dm [hunk 7]
```diff
@@ -117,6 +117,7 @@
 /atom/movable/Destroy(force)
 	QDEL_NULL(language_holder)
 	QDEL_NULL(em_block)
+	if (bound_overlay) QDEL_NULL(bound_overlay)
```
Comment (Kapu1178) [inline code/game/atoms_movable.dm:120]:
```suggestion
	if (bound_overlay) 
		QDEL_NULL(bound_overlay)
```

File: code/game/atoms_movable.dm [hunk 8]
```diff
@@ -713,6 +714,16 @@
 		else if(new_turf && !old_turf)
 			SSspatial_grid.enter_cell(src, new_turf)
 
+	// Z-Mimic hook
+	if (bound_overlay)
+		// The overlay will handle cleaning itself up on non-openspace turfs.
+		if (new_turf)
+			bound_overlay.forceMove(get_step(src, UP))
```
Comment (Kapu1178) [inline code/game/atoms_movable.dm:721]: is raw get_step() safe like this?

File: code/game/turfs/open/space/space.dm [hunk 9]
```diff
@@ -216,16 +217,7 @@ GLOBAL_REAL_VAR(starlight_color) = pick(COLOR_TEAL, COLOR_GREEN, COLOR_SILVER, C
...
-/turf/open/space/openspace/Initialize(mapload) // handle plane and layer here so that they don't cover other obs/turfs in Dream Maker
-	. = ..()
-	overlays += GLOB.openspace_backdrop_one_for_all //Special grey square for projecting backdrop darkness filter on it.
-	icon_state = "invisible"
-	return INITIALIZE_HINT_LATELOAD
-
-/turf/open/space/openspace/LateInitialize()
-	. = ..()
-	AddElement(/datum/element/turf_z_transparency, is_openspace = TRUE)
+	z_flags = Z_MIMIC_BELOW | Z_MIMIC_OVERWRITE | Z_MIMIC_NO_AO
```
Comment (Kapu1178) [inline code/game/turfs/open/space/space.dm:220]: Inherit flags from parente
```suggestion
	z_flags = Z_ATMOS_IN_DOWN | Z_ATMOS_IN_UP | Z_ATMOS_OUT_DOWN | Z_ATMOS_OUT_UP | Z_MIMIC_BELOW | Z_MIMIC_OVERWRITE | Z_MIMIC_NO_AO
```

**Review by Kapu1178 — CHANGES_REQUESTED** [review]
Small stuff

File: code/game/atoms_movable.dm (hunk 8 above)
Comment (Lohikar) [reply code/game/atoms_movable.dm:721]: It's safe, but only because of ZM's linearity requirement/assumption.

**Review by Kapu1178 — CHANGES_REQUESTED** [review]
Emissives appear to be completely borked.

**Review by Kapu1178 — CHANGES_REQUESTED** [review]
Emissives fixed (mf tg breaking mutable appearances).

Present issue: ZM is like, borked, when mimicing turfs in area/space. Very bizarre.

File: code/game/atoms.dm [hunk 10]
```diff
@@ -704,6 +704,10 @@
 	if(desc)
 		. += desc
 
+	if(user.z != z)
+		var/diff = abs(user.z - z)
```
Comment (Kapu1178) [inline code/game/atoms.dm:708]: I made an oopsie - I wasnt aware x/y/z were 0 if you aren't on a turf.
```suggestion
	if(z > 0 && user.z != z)
		var/diff = abs(user.z - z)
```

**Comment by Lohikar** [comment]
Per comment on discord, the only remaining issue is space parallax mimic as far as I know -- it's interacting poorly with render plates, so it might be best to just disable parallax mimic for now.

**Pull request merged**